### PR TITLE
CodeCamp #152 Protoseg for cityscape in mmsegmentation

### PR DESCRIPTION
**Pull request by ZephyRUXS**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This Motivation is from the article Rethinking Semantic Segmentation: A Prototype View

## Modification

Build Protoseg Decode Head in MMsegmentation structure

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Fix the error
2. Use MMCV to replace Torch api
3. Add docstring
4.

### CODE REVIEW

**Comment by xiexinch** [comment]
Hi @ZephyRUXS,
Thanks for your contribution, please sign the CLA.

File: mmseg/models/decode_heads/protoseg_head.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        self.cls_head = nn.Sequential(
+            nn.Conv2d(in_channels, in_channels, kernel_size=3, stride=1, padding=1),
+            ModuleHelper.BNReLU(in_channels, bn_type="torchbn"),
+            nn.Dropout2d(0.10)
+        )
+        
+        self.prototypes = nn.Parameter(torch.zeros(self.num_classes, self.num_prototype, in_channels),
+                                       requires_grad=True)
+
+        self.proj_head = ProjectionHead(in_channels, in_channels)
```
Comment (xiexinch) [inline mmseg/models/decode_heads/protoseg_head.py:81]: Since `ProjectionHead` is just a conv2d, the class `ProjectionHead` is not needed.

File: mmseg/models/decode_heads/protoseg_head.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        self.feat_norm = nn.LayerNorm(in_channels)
+        self.mask_norm = nn.LayerNorm(self.num_classes)
```
Comment (xiexinch) [inline mmseg/models/decode_heads/protoseg_head.py:83]: Might use the register to build the norm layer.

File: mmseg/models/decode_heads/protoseg_head.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    def prototype_learning(self, _c, out_seg, gt_seg, masks):
```
Comment (xiexinch) [inline mmseg/models/decode_heads/protoseg_head.py:84]: Might write a docstring to introduce these parameters.

File: mmseg/models/decode_heads/protoseg_head.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        pred_seg = torch.max(out_seg, 1)[1]
+        mask = (gt_seg == pred_seg.view(-1))
+        cosine_similarity = torch.mm(_c, self.prototypes.view(-1, self.prototypes.shape[-1]).t())
+        proto_logits = cosine_similarity
```
Comment (xiexinch) [inline mmseg/models/decode_heads/protoseg_head.py:88]:
```suggestion
        proto_logits = torch.mm(_c, self.prototypes.view(-1, self.prototypes.shape[-1]).t())
```

File: mmseg/models/decode_heads/protoseg_head.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                protos[k, n != 0, :] = new_value
+
+            proto_target[gt_seg == k] = indexs.float() + (self.num_prototype * k)
+
+        self.prototypes = nn.Parameter(l2_normalize(protos),
+                                           requires_grad=False)
+
+        if dist.is_available() and dist.is_initialized():
+            protos = self.prototypes.data.clone()
+            dist.all_reduce(protos.div_(dist.get_world_size()))
```
Comment (xiexinch) [inline mmseg/models/decode_heads/protoseg_head.py:130]: Can't find the definition of `dist`.

Comment (xiexinch) [reply mmseg/models/decode_heads/protoseg_head.py:130]: MMEngine implements all the distributed training logits, there is no need to implement it in the model.

File: mmseg/models/decode_heads/protoseg_head.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                ModuleHelper.BNReLU(dim_in, bn_type=bn_type),
+                nn.Conv2d(dim_in, proj_dim, kernel_size=1)
+            )
+
+    def forward(self, x):
+        return F.normalize(self.proj(x), p=2, dim=1)
+
+class Protoseg_head(nn.Module):
+    def __init__(self,in_channels):
+        super(Hrnet,self).__init__()
```
Comment (xiexinch) [inline mmseg/models/decode_heads/protoseg_head.py:65]:
```suggestion
class Protoseg_head(nn.Module):
    def __init__(self,in_channels):
        super().__init__()
```

**Review by xiexinch — CHANGES_REQUESTED** [review]
Hi @ZephyRUXS,
This method should implement based on MMSeg dev-1.x. And did you try to run a test using an official checkpoint?

File: mmseg/models/decode_heads/protoseg_head.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        elif bn_type == 'gn':
+            return nn.Sequential(
+                nn.GroupNorm(num_groups=8, num_channels=num_features, **kwargs),
+                nn.ReLU()
+            )
+        elif bn_type == 'fn':
+            exit(1)
+
+        else:
+            exit(1)
```
[the remark this answers is not in the document]
Comment (MengzhangLI) [reply mmseg/models/decode_heads/protoseg_head.py:40]: from mmcv.cnn import ConvModule

File: mmseg/models/decode_heads/protoseg_head.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        self.num_prototype=10
+        self.update_prototype=True
+        self.gamma=0.999
+        self.pretrain_prototype=False
+        self.num_classes=19
+        self.cls_head = nn.Sequential(
+            nn.Conv2d(in_channels, in_channels, kernel_size=3, stride=1, padding=1),
+            ModuleHelper.BNReLU(in_channels, bn_type="torchbn"),
+            nn.Dropout2d(0.10)
+        )
```
[the remark this answers is not in the document]
Comment (ZephyRUXS) [reply mmseg/models/decode_heads/protoseg_head.py]: This fix

> These properties should support user customizing.

File: mmseg/models/decode_heads/protoseg_head.py (hunk 4 above)
Comment (ZephyRUXS) [reply mmseg/models/decode_heads/protoseg_head.py]: > 

This fix

File: mmseg/models/decode_heads/protoseg_head.py (hunk 5 above)
Comment (ZephyRUXS) [reply mmseg/models/decode_heads/protoseg_head.py:145]: > Can't find the definition of `dist`.
This fix

File: mmseg/models/decode_heads/protoseg_head.py (hunk 6 above)
Comment (ZephyRUXS) [reply mmseg/models/decode_heads/protoseg_head.py]: > 

This fix
